File: synara/src/app/features/matrix-dto/member.ts
/**
 * Room member DTO.
 */

import type { RoomId, UserId } from './ids';
import { isMembership, type Membership } from './room';
import {
  hasForbiddenWireFields,
  isObject,
  optBoolean,
  optString,
  reqNumber,
  reqString,
} from './parseUtil';

export type RoomMember = {
  roomId: RoomId;
  userId: UserId;
  displayName?: string;
  avatarUrl?: string;
  membership: Membership;
  powerLevel: number;
  isDirectTarget?: boolean;
};

export function parseRoomMember(value: unknown): RoomMember | null {
  if (!isObject(value) || hasForbiddenWireFields(value)) return null;
  const roomId = reqString(value, 'roomId');
  const userId = reqString(value, 'userId');
  const displayName = optString(value, 'displayName');
  const avatarUrl = optString(value, 'avatarUrl');
  const powerLevel = reqNumber(value, 'powerLevel');
  const isDirectTarget = optBoolean(value, 'isDirectTarget');
  if (
    roomId === null ||
    userId === null ||
    displayName === null ||
    avatarUrl === null ||
    powerLevel === null ||
    isDirectTarget === null ||
    !isMembership(value.membership)
  ) {
    return null;
  }
  return {
    roomId,
    userId,
    displayName,
    avatarUrl,
    membership: value.membership,
    powerLevel,
    isDirectTarget,
  };
}

File: crates/synara-core/src/dto/member.rs
//! Room member DTO — list/drawer projection.

use serde::{Deserialize, Serialize};

use super::ids::{RoomId, UserId};
use super::room::Membership;

/// Single room member projection.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RoomMember {
    pub room_id: RoomId,
    pub user_id: UserId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
    /// mxc or product media-handle URI — string only.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub avatar_url: Option<String>,
    pub membership: Membership,
    pub power_level: i32,
    /// True when this member is the DM peer for a direct room.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub is_direct_target: Option<bool>,
}
